pages/chaoss/visualizations/contrib_importance_over_time.py

```diff
@@ -0,0 +1,416 @@
+                                        "About Graph",
+                                        id=f"popover-target-{PAGE}-{VIZ_ID}",
+                                        color="secondary",
+                                        size="sm",
+                                    ), ],
+                                    width="auto",
+                                    style={"paddingTop": ".5em"},
+                                ),
+                            ], 
+                            align="center",
+                            justify="between",
+                        ), 
+                    ]
+                ),
+            ]
+        )
+    ],
+)
+
+# callback for graph info popover
+@callback(
+    Output(f"popover-{PAGE}-{VIZ_ID}", "is_open"),
+    [Input(f"popover-target-{PAGE}-{VIZ_ID}", "n_clicks")],
+    [State(f"popover-{PAGE}-{VIZ_ID}", "is_open")],
+)
+def toggle_popover(n, is_open):
+    if n:
+        return not is_open
+    return is_open
+
+# callback for dynamically changing the graph title
+@callback(
+    Output(f"graph-title-{PAGE}-{VIZ_ID}", "children"),
+    Input(f"window-width-{PAGE}-{VIZ_ID}", "value"),
+)
+
+def graph_title(window_width):
+    title = f"Contributor Importance in {window_width} Month Windows"
+    return title
+
+# callback for contrib-importance-over-time graph
+@callback(
+    Output(f"{PAGE}-{VIZ_ID}", "figure"),
+    Output(f"check-alert-{PAGE}-{VIZ_ID}", "is_open"),
+    [   
+        Input("repo-choices", "data"),
+        Input(f"patterns-{PAGE}-{VIZ_ID}", "value"),
+        Input(f"threshold-{PAGE}-{VIZ_ID}", "value"),
+        Input(f"window-width-{PAGE}-{VIZ_ID}", "value"),
+        Input(f"step-size-{PAGE}-{VIZ_ID}", "value"),
+        Input(f"date-picker-range-{PAGE}-{VIZ_ID}", "start_date"),
+        Input(f"date-picker-range-{PAGE}-{VIZ_ID}", "end_date"),
+    ],
+    background=True,
+)
+
+def create_contrib_importance_over_time_graph(repolist, patterns, threshold, window_width, step_size, start_date, end_date):
+    # main function for all data pre processing
+    cache = cm()
+    df = cache.grabm(func=ctq, repos=repolist)
+    
+    while df is None:
+        time.sleep(1.0)
+        df = cache.grabm(func=ctq, repos=repolist)
+
+    # data ready.
+    start = time.perf_counter()
+    logging.warning(f"{VIZ_ID}- START")
+
+    # test if there is data
+    if df.empty:
+        logging.warning(f"{VIZ_ID} - NO DATA AVAILABLE")
+        return nodata_graph
+    
+    # if the step size is greater than window width raise Alert
+    if step_size > window_width:
+        return dash.no_update, True
+    
+    df_final = process_data(df, 
+                            patterns, 
+                            threshold, 
+                            window_width, 
+                            step_size, 
+                            start_date, 
+                            end_date)
+    
+    # drop rows with NaN values 
+    df_final = df_final.dropna(axis=0)
```

Comment:
Why drop rows with NAN values? What condition creates rows with NAN values?

Reply:
There are alot of NaN values for contributor importance observed during the start of the project which makes sense. However, for commits, sometimes contributors aren't identified i.e have None for their cntrb_id so we can't include their data in contributor importance. Should I drop rows with cntrb_id = None in process_data instead of dropping rows in df_final ?

Reply:
Ill be curious to hear what @JamesKunstle says, I think it should be done at the beginning of calc_cntrb_importance . The reason for my thought of this is for the cntrb_importance is not skewed by all the contributors with NAN as an ID getting considered as 1 contributor. Luckily this should only impact commits